File: Programmers/programmers_문자열압축.cpp
#include <string>
#include <vector>
#include <cstring>
#include <iostream>
#include <cstdio>
using namespace std;

int solution(string s) {
    int answer = 1000000000;
    int cut = 1;
    if (s.length() == 1) return 1;
    while (cut <= (int)(s.length() / 2)) {
        string cmpStr = s.substr(0, cut);
        //cout << cmpStr << endl;
        int cnt = 0;
        int len = 0;
        //cout << "len:"<<len << "\n";
        for (int i = 0; i < s.length(); i += cut) {
            if (i + cut > s.length()) {
                len += s.length() - i;
                break;
            }
            string curStr = s.substr(i, cut);
            //cout << "curcomStr:" << cmpStr << "\t";
            //cout << "curStr:" << curStr << endl;
            if (curStr == cmpStr) {
                cnt++;
            }
            else {
                if (cnt != 1) {
                    //len++;
                    len+=to_string(cnt).length();
                }
                len += cmpStr.length();
                cnt = 1;
                cmpStr.clear();
                cmpStr = curStr;
            }
            //cout << "curlen:" << len << endl;
        }
        if (cnt != 1) {
            len += cmpStr.length() + to_string(cnt).length();
        }
        else {
            len += cmpStr.length();
        }
        //cout << len << endl;
        cut++;
        if (answer > len) answer = len;
    }
    return answer;
}
int main() {
    string str;
    cin >> str;
    cout << solution(str);
}

File: Programmers/programmers_문자열 압축.py
def chkLen(length, s):
    res = 0
    substr = s[0:length]
    same = 1
    for sidx in range(length, len(s), length):
        if substr == s[sidx:sidx+length] :
            same += 1
            continue
        if same == 1:
            res += length
        else :
            res += length + len(str(same))
        substr = s[sidx:sidx+length]
        same = 1
    if same > 1:
        res += len(substr) + len(str(same))
    else :
        res += len(substr)
    return res
    
def solution(s):
    if len(s) == 1 : return 1
    answer = 1001
    #1개단위~문자열 길이/2 만큼 자르는 단위를 각각 검사
    tmpAns = 0
    for length in range(1,int(len(s)/2)+1):
        tmpAns = chkLen(length, s)
        if tmpAns < answer :
            answer = tmpAns
    return answer
